src/Orleans.Core/Networking/ConnectionManager.cs

```diff
@@ -53,158 +55,162 @@ public int ConnectionCount
-                    await Task.Delay(10);
-                }
-
-                // Attempt to connect.
-                Connection connection = default;
-                try
-                {
-                    connection = await this.ConnectAsync(endpoint);
-                    entry = OnConnectedInternal(endpoint, connection);
-                    result = entry.Connections;
-                }
-                catch (Exception exception)
-                {
-                    OnConnectionFailed(endpoint, DateTime.UtcNow);
-                    throw new ConnectionFailedException(
-                        $"Unable to connect to endpoint {endpoint}. See {nameof(exception.InnerException)}", exception);
                 }
             }
-            finally
-            {
-                if (acquiredConnectionLock) entry.ReleaseLock();
-            }
 
-            nextConnection = (nextConnection + 1) % result.Length;
-            return result[nextConnection];
+            return new ValueTask<Connection>(connection);
         }
 
-        private void OnConnectionFailed(SiloAddress address, DateTime lastFailure)
+        private async Task<Connection> GetConnectionAsync(SiloAddress endpoint)
         {
-            bool acquiredConnectionLock = false;
-            ConnectionEntry entry = default;
-            lock (this.collectionLock)
+            while (true)
             {
-                try
+                var entry = this.GetOrCreateEntry(endpoint);
+                if (entry.Connections.Length >= this.connectionOptions.ConnectionsPerEndpoint
+                    && entry.TryGetNextConnection(out var connection))
+                {
+                    return connection;
+                }
+
+                // Wait for the pending attempt to complete
+                var pendingTask = entry.PendingConnection;
+                if (pendingTask != null && !pendingTask.IsCompleted)
                 {
-                    entry = this.GetOrCreateEntry(address, ref acquiredConnectionLock);
+                    await pendingTask.ConfigureAwait(false);
+                }
+
+                entry = this.GetOrCreateEntry(endpoint);
+                entry.ThrowIfRecentFailure(endpoint, this.connectionOptions.ConnectionRetryDelay);
 
-                    if (entry.LastFailure.HasValue)
+                // Attempt to connect.                
+                lock (this.lockObj)
+                {
+                    pendingTask = entry.PendingConnection;
+                    if (pendingTask == null || pendingTask.IsCompleted)
                     {
-                        var ticks = Math.Max(lastFailure.Ticks, entry.LastFailure.Value.Ticks);
-                        lastFailure = new DateTime(ticks);
+                        pendingTask = this.ConnectAsync(endpoint);
+                        entry.PendingConnection = pendingTask;
                     }
+                }
```

Comment:
Note that we *start* the async connection attempt under the lock and await it outside of the lock. This is to ensure that we correctly set `PendingConnection` and don't have a race for multiple pending connections. Another alternative would be to use a `TaskCompletionSource` which could be assigned from within the lock but completed in the background by `ConnectAsync`